core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -308,6 +311,15 @@ object SparkSubmit extends CommandLineUtils {
       RPackageUtils.checkAndBuildRPackage(args.jars, printStream, args.verbose)
     }
 
+    // In client mode, download remote files.
+    if (deployMode == CLIENT) {
```

Comment:
It seems it can handle remote files in Yarn/Mesos cluster mode. I haven't tested it, because we are using client mode.